server/api.js

```diff
@@ -0,0 +1,551 @@
+
+async function handleGetClientGateways(req, res) {
+  const validationErrors = Validator.validationResult(req).formatWith(helpers.formatExpressValidationErrors)
+
+  try {
+    if (validationErrors.isEmpty()) {
+      const gateways = await db.getGatewaysWithClientId(req.params.clientId)
+
+      // if the query failed and returned null, the clientId is probably wrong
+      if (gateways == null) {
+        res.status(404).send({ status: 'error', message: 'Not Found' })
+
+        return
+      }
+
+      res.status(200).send({ status: 'success', data: gateways })
+    } else {
+      res.status(400).send({ status: 'error', message: 'Bad Request' })
+      helpers.logError(`Bad request to ${req.path}: ${validationErrors.array()}`)
+    }
+  } catch (error) {
+    res.status(500).send({ status: 'error', message: 'Internal Server Error' })
+    helpers.logError(`Internal server error at ${req.path}: ${error.message}`)
+  }
+}
+
+const validateGetClientVitals = [Validator.header(['Authorization']).notEmpty(), Validator.param(['clientId']).notEmpty()]
+
+async function handleGetClientVitals(req, res) {
+  const validationErrors = Validator.validationResult(req).formatWith(helpers.formatExpressValidationErrors)
+
+  try {
+    if (validationErrors.isEmpty()) {
+      const buttonVitals = await db.getRecentButtonsVitalsWithClientId(req.params.clientId)
+      const gatewayVitals = await db.getRecentGatewaysVitalsWithClientId(req.params.clientId)
+
+      // if either of the query failed and returned null, the clientId is probably wrong
+      if (buttonVitals == null || gatewayVitals == null) {
+        res.status(404).send({ status: 'error', message: 'Not Found' })
+
+        return
+      }
+
+      res.status(200).send({ status: 'success', data: { buttonVitals, gatewayVitals } })
+    } else {
+      res.status(400).send({ status: 'error', message: 'Bad Request' })
+      helpers.logError(`Bad request to ${req.path}: ${validationErrors.array()}`)
+    }
+  } catch (error) {
+    res.status(500).send({ status: 'error', message: 'Internal Server Error' })
+    helpers.logError(`Internal server error at ${req.path}: ${error.message}`)
+  }
+}
+
+const validateRegisterClient = [
+  Validator.header(['Authorization']).notEmpty(),
+  Validator.body(['displayName', 'fromPhoneNumber', 'language']).trim().isString().notEmpty(),
+  Validator.body(['responderPhoneNumbers', 'fallbackPhoneNumbers', 'heartbeatPhoneNumbers', 'incidentCategories']).isArray(),
+  Validator.body(['reminderTimeout', 'fallbackTimeout']).isInt({ min: 0 }),
+  Validator.body(['isDisplayed', 'isSendingAlerts', 'isSendingVitals']).isBoolean(),
+]
+
+async function handleRegisterClient(req, res) {
+  const validationErrors = Validator.validationResult(req).formatWith(helpers.formatExpressValidationErrors)
+
+  try {
+    if (validationErrors.isEmpty()) {
+      const client = await db.createClient(
+        req.body.displayName,
+        req.body.responderPhoneNumbers,
```

Comment:
I'm going to leave this open since I'm not sure how to use Validator to validate the individual items of a submitted Array.

I'd like to validate:
- phone numbers with regex after trim
- incident categories after trim

Reply:
I think you'd need to write a custom validator for it. Not sure if it's worth it. 

https://express-validator.github.io/docs/6.5.0/custom-validators-sanitizers/